### PR TITLE
Implement websockets based on-demand-entries ping

### LINKED ISSUE #4495: Replace `on-demand-retries` XHR ping with websocket
During development, the `on-demand-retries` ping causes a lot of entries in the devtools, which make debugging a bit more annoying than it should be.

Switching to WebSocket would give us a nice performance boost as well.

https://github.com/zeit/next.js/issues/1364

### PR DESCRIPTION
**Pull request by ijjk**
Fixes #4495

Here's my approach for replacing the XHR on-demand-entries pinger #1364 #4495. I'm not sure if this is the way everyone wants to accomplish this since I saw mention of using a separate server and port for the dynamic entries websocket, but thought this would be a fairly clean solution since it doesn't need that.

With this method the only change when using a custom server is you have to listen for the upgrade event and pass it to next.getRequestHandler(). Example: 
```
const server = app.listen(port)
const handleRequest = next.getRequestHandler()

if(dev) {
  server.on('upgrade', handleRequest)
}
```

### CODE REVIEW

**Comment by ijjk** [comment]
Hey @rauchg  I was wondering if I could get your feedback on this. Sorry to bother if you're the wrong person to ask, I thought I'd ask you since you prioritized the issue.

**Comment by ijjk** [comment]
Okay cool, thanks for the reply!

**Comment by willopez** [comment]
@timneutkens can this be revisited?

**Comment by Enalmada** [comment]
Having to put a little code in custom server seems worth it.  Are there any other disadvantages of this solution over the current one?

**Comment by Enalmada** [comment]
Would it make reviewers more comfortable if this was configurable and people could opt-in to websocket based pinger?  I would happily beta test this feature if there is some hesitation about making it the default right away.

**Comment by ijjk** [comment]
Hey @Enalmada I updated my [dynamic-entries-websocket](https://github.com/ijjk/next.js/tree/dynamic-entries-websocket) branch with the new next structure so if you want to give it a try it should be good to go.

**Comment by Enalmada** [comment]
@ijjk Thanks for taking the time to do this pull request. It seems to work great!  Developers like me that spend lots of time with the chrome console open will be so happy when this is merged.  It is a superior developer experience and has tests and everything.     

BTW Can you elaborate on what `stripBase` is and why?  I removed it from my local test just to try to get things working:
```
    if (dev) {
        server.on("upgrade", (req, socket, head) => {
            handle(req, socket, parse(req.url, true));
        });
    }
```

**Comment by timneutkens** [comment]
Is there an alternative to doing this:

```
if(dev) {
  server.on('upgrade', handleRequest)
}
```

I assume we could open a socket server on a different (random) port in development?

This is the only part that makes me hesitate to merge this PR, as it'll break pretty much every implementation of Next.js that is currently out there, and it's easy to forget adding this.

**Comment by ijjk** [comment]
Yeah, that makes sense, it does seem better to avoid that. I updated my branch to create a WebSocket server on a random port for the dynamic-entries pinging in development.

**Comment by timneutkens** [comment]
💯 Going to have a look 👍

File: packages/next/server/on-demand-entry-handler.js [hunk 1]
```diff
@@ -45,6 +43,11 @@ export default function onDemandEntryHandler (devMiddleware, multiCompiler, {
   let reloading = false
   let stopped = false
   let reloadCallbacks = new EventEmitter()
+  let port = null
+  // create websocket on random port for dynamic entries pinging
+  const wss = new WebSocket.Server({ port: 0 }, () => {
```
Comment (timneutkens) [inline packages/next/server/on-demand-entry-handler.js]: If you create the websocket server here: https://github.com/zeit/next.js/blob/canary/packages/next/server/hot-reloader.js#L167

And wait till the port is there, you'll be able to pass the port to the webpack config, then in the webpack config you can use `DefinePlugin` to set `process.env.NEXT_WS_PORT` or similar to the generated port. And use that in the client side code to read the value, that way you don't have to do the extra fetch on page load just to get the port 👍

Comment (ijjk) [reply packages/next/server/on-demand-entry-handler.js]: Good idea, I moved it there and added the `DefinePlugin` to the client webpack config

**Review by timneutkens — APPROVED** [review]
Let's put this on canary after my changes to render.js 💯

**Comment by glenarama** [comment]
Is there a way to manually set the port? I have to manually map ports so this has corrupted my Dev experience. I did try setting process.env.NEXT_WS_PORT on both client and server.

**Comment by ijjk** [comment]
@glenarama currently you can't manually set the port for the dynamic-entries WebSocket. I just sent over a [pull request](https://github.com/zeit/next.js/pull/5963) adding a `--websocket` or `-w` argument similar to how the `-p` argument works to fix that.

**Comment by glenarama** [comment]
Great, thanks!

**Pull request merged**